src/ar/Tag.cpp

```diff
@@ -0,0 +1,151 @@
+#include "Tag.h"
+
+namespace AR
+{
+	double findAngle(cv::Point a, cv::Point b, cv::Point c)
+	{
+		a = a - b; // shift vectors so that B is at the origin
+		b = b - b;
+		c = c - b;
+		double dot_product = a.ddot(c);
+		double a_mag = sqrt((a.x * a.x) + (a.y * a.y));
+		double c_mag = sqrt((c.x * c.x) + (c.y * c.y));
+		return acos(dot_product / (a_mag * c_mag));
+	}
+
+	cv::Vec3d rotationMatrixToEulerAngles(cv::Mat& R)
+	{
+		double sy =
+			sqrt(R.at<double>(0, 0) * R.at<double>(0, 0) + R.at<double>(1, 0) * R.at<double>(1, 0));
+
+		bool singular = sy < 1e-6;
+
+		double x, y, z;
+		if(!singular)
+		{
+			x = atan2(R.at<double>(2, 1), R.at<double>(2, 2));
+			y = atan2(-R.at<double>(2, 0), sy);
+			z = atan2(R.at<double>(1, 0), R.at<double>(0, 0));
+		}
+		else
+		{
+			x = atan2(-R.at<double>(1, 2), R.at<double>(1, 1));
+			y = atan2(-R.at<double>(2, 0), sy);
+			z = 0;
+		}
+		return cv::Vec3f(x, y, z);
+	}
+
+	const char* InvalidCornerException::what() const throw()
```

Comment:
We should use error codes

Reply:
I will add stuff to slite about this later

Reply:
Okay. Unsure about how that will work because this exception is thrown from a constructor, in the event that the corners are in the wrong order. In this case we thought it was better to fail rather than sort the coordinates into the correct order without telling the user.

Maybe we could have an error code parameter passed by reference or something.

Reply:
Umm how do you guys plan to handle this exception when one is thrown? Currently this exception is not caught.